lib/SPIRVOp.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright 2019 The Clspv Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "SPIRVOp.h"
+
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/DerivedTypes.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+
+#include "Constants.h"
+
+namespace clspv {
+
+using namespace llvm;
+
+// The mangling loosely follows the Itanium convention.
+// Its purpose is solely to ensure uniqueness of names, it is not
+// meant to convey type information.
+static std::string mangleType(Type* Ty) {
+  if (Ty->isVectorTy()) {
+    auto NumVecElems = std::to_string(Ty->getVectorNumElements());
+    return "Dv" + NumVecElems + "_" + mangleType(Ty->getScalarType());
+  } else if (Ty->isIntegerTy()) {
+    if (Ty->getScalarSizeInBits() == 8) {
+      return "h";
+    } else if (Ty->getScalarSizeInBits() == 16) {
+      return "t";
+    } else if (Ty->getScalarSizeInBits() == 32) {
+      return "j";
+    } else if (Ty->getScalarSizeInBits() == 64) {
+      return "m";
+    }
+  }
+
+  llvm_unreachable("Unhandled type in SPIR-V intrinsic name mangler");
+}
+
+Instruction* InsertSPIRVOp(Instruction* Insert, spv::Op Opcode, Type* RetType,
+                           ArrayRef<Value*> Args) {
+
+  // Prepare mangled name
+  std::string MangledName = clspv::SPIRVOpIntrinsicFunction();
+  MangledName += std::to_string(Opcode);
+  MangledName += ".";
+  for (auto Arg : Args) {
+    MangledName += mangleType(Arg->getType());
+  }
+
+  // Create a function in the module
+  auto M = Insert->getModule();
+  auto Int32Ty = Type::getInt32Ty(M->getContext());
+  SmallVector<Type*, 8> ArgTypes = { Int32Ty };
+  for (auto Arg : Args) {
+    ArgTypes.push_back(Arg->getType());
+  }
+  auto NewFType = FunctionType::get(RetType, ArgTypes, false);
+  auto NewF = M->getOrInsertFunction(MangledName, NewFType);
```

Comment:
For the mad_hi case, it might help optimization to call NewF->setDoesNotAccessMemory() or equivalent.

I think the best way to do that is to add another parameter which is the list of attributes to be applied to the function.  For mad_hi it would include Attribute::ReadNone

Reply:
To be clear, I mean another parameter to the InsertSpirvOp function, could be an ArrayRef<Attribute>

Reply:
Yes, I shouldn't have been able to write the tests the way I did :). I was planning to look into this later. I was wondering whether we could always add this attribute based on the absence of pointer arguments. For now, I just went with your suggestion.

I've rebased the change on top of the clang/llvm update. PTAL.